File: content/examples/other/leetcode/problems/lowest-common-ancestor-of-bst/example-1/solution.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function lowestCommonAncestorBrute(root, p, q) {
  function pathTo(node, target, path) {
    if (!node) return false;
    path.push(node);
    if (node === target) return true;
    if (pathTo(node.left, target, path) || pathTo(node.right, target, path)) return true;
    path.pop();
    return false;
  }

  const pPath = [];
  const qPath = [];
  pathTo(root, p, pPath);
  pathTo(root, q, qPath);

  let i = 0;
  while (i < pPath.length && i < qPath.length && pPath[i] === qPath[i]) i += 1;
  return pPath[i - 1] || null;
}

if (require.main === module) {
  const root = new TreeNode(6);
  root.left = new TreeNode(2, new TreeNode(0), new TreeNode(4, new TreeNode(3), new TreeNode(5)));
  root.right = new TreeNode(8, new TreeNode(7), new TreeNode(9));
  const p = root.left;
  const q = root.right;
  console.log(lowestCommonAncestorBrute(root, p, q).val);
}

module.exports = { TreeNode, lowestCommonAncestorBrute };

File: content/examples/other/leetcode/problems/lowest-common-ancestor-of-bst/example-1/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def lowest_common_ancestor_brute(root, p, q):
    def path_to(node, target, path):
        if node is None:
            return False
        path.append(node)
        if node is target:
            return True
        if path_to(node.left, target, path) or path_to(node.right, target, path):
            return True
        path.pop()
        return False

    p_path = []
    q_path = []
    path_to(root, p, p_path)
    path_to(root, q, q_path)

    i = 0
    while i < len(p_path) and i < len(q_path) and p_path[i] is q_path[i]:
        i += 1
    return p_path[i - 1] if i > 0 else None


if __name__ == "__main__":
    root = TreeNode(6)
    root.left = TreeNode(2, TreeNode(0), TreeNode(4, TreeNode(3), TreeNode(5)))
    root.right = TreeNode(8, TreeNode(7), TreeNode(9))
    p = root.left
    q = root.right
    print(lowest_common_ancestor_brute(root, p, q).val)
